benchmarks/bm/_scenario.py

```diff
@@ -6,18 +6,49 @@
 import six
 
 
-def register(cls):
-    """Decorator for scenario class that registers the scenario to be run."""
-    wrapped_cls = attr.s(cls)
-    _register(wrapped_cls)
-    return wrapped_cls
+var = attr.ib
 
 
-var = attr.ib
+def _register(scenario_cls):
+    """Registers a scenario for benchmarking."""
+    # This extends pyperf's runner by registering arguments for the scenario config
+    def add_cmdline_args(cmd, args):
+        for field in attr.fields(scenario_cls):
+            if hasattr(args, field.name):
+                cmd.extend(("--{}".format(field.name), str(getattr(args, field.name))))
+
+    runner = pyperf.Runner(add_cmdline_args=add_cmdline_args)
+    cmd = runner.argparser
+
+    for field in attr.fields(scenario_cls):
+        cmd.add_argument("--{}".format(field.name), type=field.type)
+
+    parsed_args = runner.parse_args()
+
+    config_dict = {
+        field.name: getattr(parsed_args, field.name)
+        for field in attr.fields(scenario_cls)
+        if hasattr(parsed_args, field.name)
+    }
+    scenario = scenario_cls(**config_dict)
+
+    runner.bench_time_func(scenario.scenario_name, scenario._pyperf)
+
+
+class ScenarioMeta(abc.ABCMeta):
+    def __init__(cls, name, bases, _dict):
+        super(ScenarioMeta, cls).__init__(name, bases, _dict)
 
+        # Make sure every sub-class is wrapped by `attr.s`
+        cls = attr.s()(cls)
 
-@attr.s
-class Scenario(six.with_metaclass(abc.ABCMeta)):
+        # Do not register the base Scenario class
+        # DEV: We cannot compare `cls` to `Scenario` since it doesn't exist yet
+        if cls.__module__ != __name__:
+            _register(cls)
```

Comment:
Can the check be moved inside `_register` so that `Scenario` exists?

Reply:
it still won't exist until after this whole function exists, but we can move this `cls.__module__ != __name__` check to `_register`, or else we can do a name check like `cls.__name__ != "Scenario"`, but that seemed a little more error prone if someone did `class Scenario(bm.Scenario)`.